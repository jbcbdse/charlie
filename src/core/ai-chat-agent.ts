import {
  ChatExecutor,
  ChatAgentGetResponseInput,
  ChatAgentGetResponseOutput,
  ChatMessageTransformer,
  ChatAgent,
  MessageToolCall,
  ChatAgentContext,
  MessageTool,
} from "./types";
import { ToolExecutor } from "./tool-executor";
import { EventName, eventProducer, EventProducer } from "./event-producer";
import { newRunId } from "./new-run-id";
import { BaseTool } from "./base-tool";
import { TemplateSerializer } from "./template-serializer";

export class AiChatAgent implements ChatAgent {
  private chatExecutor: ChatExecutor;
  private toolExecutor: ToolExecutor;
  private preToolCallTransformers: ChatMessageTransformer[] = [];
  private postToolCallTransformers: ChatMessageTransformer[] = [];
  private postRunTransformers: ChatMessageTransformer[] = [];
  private eventProducer: EventProducer;
  private systemPromptTemplate?: string;
  private templateSerializer: TemplateSerializer;
  constructor(options: {
    chatExecutor: ChatExecutor;
    toolExecutor?: ToolExecutor;
    systemPromptTemplate?: string;
    templateSerializer?: TemplateSerializer;
    preToolCallTransformers?: ChatMessageTransformer[];
    postToolCallTransformers?: ChatMessageTransformer[];
    postRunTransformers?: ChatMessageTransformer[];
  }) {
    this.chatExecutor = options.chatExecutor;
    this.toolExecutor = options.toolExecutor || new ToolExecutor();
    this.preToolCallTransformers = options.preToolCallTransformers || [];
    this.postToolCallTransformers = options.postToolCallTransformers || [];
    this.postRunTransformers = options.postRunTransformers || [];
    this.eventProducer = eventProducer;
    this.systemPromptTemplate = options.systemPromptTemplate;
    this.templateSerializer =
      options.templateSerializer || new TemplateSerializer();
  }
  public async getResponse({
    messages,
    tools,
    meta = {},
  }: ChatAgentGetResponseInput): Promise<ChatAgentGetResponseOutput> {
    const systemPrompt = this.systemPromptTemplate
      ? this.templateSerializer.serialize(this.systemPromptTemplate, meta)
      : undefined;
    let doLoop = true;
    let responseMessages: ChatAgentGetResponseOutput["responseMessages"] = [];
    const context: ChatAgentContext = {
      runId: newRunId(),
      modelId: this.chatExecutor.modelId,
      messages,
      meta,
    };
    context.runId = newRunId();
    context.modelId = this.chatExecutor.modelId;
    context.messages = messages;
    const chatStartMs = Date.now();
    this.eventProducer.emit(EventName.ChatStart, {
      context,
      messages,
      systemPrompt,
      modelId: this.chatExecutor.modelId,
    });
    do {
      const chatExecutorStartMs = Date.now();
      this.eventProducer.emit(EventName.ChatExecutorStart, {
        context,
        messages,
        modelId: this.chatExecutor.modelId,
      });
      const response = await this.chatExecutor.execute({
        messages,
        tools,
        systemPrompt,
        context,
      });
      let newResponseMessages = response.responseMessages;
      for (const transformer of this.preToolCallTransformers) {
        newResponseMessages = await transformer.transform(newResponseMessages);
      }
      responseMessages.push(...newResponseMessages);
      this.eventProducer.emit(EventName.ChatExecutorEnd, {
        context,
        messages: newResponseMessages,
        modelId: this.chatExecutor.modelId,
        timeMs: Date.now() - chatExecutorStartMs,
      });
      const toolCalls = newResponseMessages.filter(
        (m) => m.role === "tool_call",
      );
      if (tools && toolCalls.length > 0) {
        const toolResponses = await this.executeToolCalls(
          toolCalls,
          tools,
          context,
        );
        responseMessages.push(...toolResponses);
        let newMessages = [...newResponseMessages, ...toolResponses];
        for (const transformer of this.postToolCallTransformers) {
          newMessages = await transformer.transform(newMessages);
        }
        messages = [...messages, ...newMessages];
      } else {
        doLoop = false;
      }
    } while (doLoop);
    for (const transformer of this.postRunTransformers) {
      responseMessages = await transformer.transform(responseMessages);
    }
    const response = {
      responseMessage: responseMessages[responseMessages.length - 1],
      responseMessages,
    };
    this.eventProducer.emit(EventName.ChatEnd, {
      context,
      messages: response.responseMessages,
      modelId: this.chatExecutor.modelId,
      timeMs: Date.now() - chatStartMs,
    });
    return response;
  }

  private async executeToolCalls(
    toolCalls: MessageToolCall[],
    tools: BaseTool[],
    context: ChatAgentContext,
  ): Promise<MessageTool[]> {
    const toolStartMs = Date.now();
    this.eventProducer.emit(EventName.ToolsStart, {
      context,
      toolCalls,
    });
    const toolMessages = await Promise.all(
      toolCalls.map((toolCall) =>
        this.toolExecutor.execute(toolCall, tools, context),
      ),
    ).then((results) => results.flat());
    this.eventProducer.emit(EventName.ToolsEnd, {
      context,
      toolMessages,
      timeMs: Date.now() - toolStartMs,
    });
    return toolMessages;
  }
}
