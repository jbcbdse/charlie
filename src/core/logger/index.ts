import winston from "winston";

export interface ILogger {
  error: (message: string, meta?: Record<string, unknown>) => void;
  warn: (message: string, meta?: Record<string, unknown>) => void;
  info: (message: string, meta?: Record<string, unknown>) => void;
  verbose: (message: string, meta?: Record<string, unknown>) => void;
  debug: (message: string, meta?: Record<string, unknown>) => void;
}
type LogLevel = "error" | "warn" | "info" | "verbose" | "debug";
export class Logger implements ILogger {
  private logLevel: LogLevel;
  private logger: winston.Logger;
  constructor(options: { logLevel?: LogLevel } = {}) {
    this.logLevel = options.logLevel || "debug";
    this.logger = winston.createLogger({
      level: this.logLevel,
      format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.colorize(),
        winston.format.simple(),
      ),
      transports: [
        new winston.transports.Console({
          stderrLevels: ["error", "warn", "info", "verbose", "debug"],
        }),
      ],
    });
  }
  public error(message: string, meta?: Record<string, unknown>) {
    this.logger.error(message, meta);
  }
  public warn(message: string, meta?: Record<string, unknown>) {
    this.logger.warn(message, meta);
  }
  public info(message: string, meta?: Record<string, unknown>) {
    this.logger.info(message, meta);
  }
  public verbose(message: string, meta?: Record<string, unknown>) {
    this.logger.verbose(message, meta);
  }
  public debug(message: string, meta?: Record<string, unknown>) {
    this.logger.debug(message, meta);
  }
}
