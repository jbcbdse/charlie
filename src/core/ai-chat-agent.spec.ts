import { CalculatorTool } from "../tools/calculator.tool";
import { AiChatAgent } from "./ai-chat-agent";
import { ILogger } from "./logger";
import {
  ChatAgentGetResponseInput,
  ChatAgentGetResponseOutput,
  ChatExecutor,
  ChatMessage,
  MessageAssistant,
} from "./types";
import { mock } from "jest-mock-extended";

class MockExecutor implements ChatExecutor {
  modelId = "mock-model-id";
  execute = jest.fn(function (
    input: ChatAgentGetResponseInput,
  ): Promise<ChatAgentGetResponseOutput> {
    const message = input.messages[input.messages.length - 1];
    let responseMessage: ChatMessage = { role: "assistant", content: "Hello" };
    // mock what the LLM does when the user asks a question that requires a tool
    if (message.role === "user" && message.content === "Calculate 3 + 4") {
      const expr = message.content.replace("Calculate ", "");
      responseMessage = {
        role: "tool_call",
        toolCalls: [
          {
            function: {
              name: "CalculatorTool",
              arguments: { expr },
            },
            id: "toolcall1",
            type: "function",
          },
        ],
      };
    } else if (message.role === "user") {
      responseMessage = {
        role: "assistant",
        content: `You said: ${message.content}`,
      };
    }
    if (message.role === "tool") {
      responseMessage = {
        role: "assistant",
        content: `The ${message.name} tool said: ${message.content}`,
      };
    }
    return Promise.resolve({
      responseMessage,
      responseMessages: [responseMessage],
    });
  });
}

describe("AiChatAgent", () => {
  let agent: AiChatAgent;
  let chatExecutor: ChatExecutor;
  beforeEach(() => {
    chatExecutor = new MockExecutor();
    agent = new AiChatAgent({
      chatExecutor,
      logger: mock<ILogger>(),
    });
  });
  describe("getResponse", () => {
    it("should return a response message", async () => {
      const response = await agent.getResponse({
        context: {},
        messages: [{ role: "user", content: "Hey buddy" }],
        tools: [new CalculatorTool()],
      });
      const responseMessage = response.responseMessage as MessageAssistant;
      expect(responseMessage.role).toBe("assistant");
      expect(responseMessage.content).toBe("You said: Hey buddy");
    });
    it("should call a tool and return the response, along with intermediate steps including tool calls", async () => {
      const response = await agent.getResponse({
        context: {},
        messages: [{ role: "user", content: "Calculate 3 + 4" }],
        tools: [new CalculatorTool()],
      });
      const responseMessage = response.responseMessage as MessageAssistant;
      expect(responseMessage.role).toBe("assistant");
      expect(responseMessage.content).toBe("The CalculatorTool tool said: 7");
      expect(response.responseMessages).toMatchSnapshot();
    });
  });
});
