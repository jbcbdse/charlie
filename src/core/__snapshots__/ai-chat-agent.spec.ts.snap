// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AiChatAgent getResponse should call a tool and return the response, along with intermediate steps including tool calls 1`] = `
[
  {
    "role": "tool_call",
    "toolCalls": [
      {
        "function": {
          "arguments": {
            "expr": "3 + 4",
          },
          "name": "CalculatorTool",
        },
        "id": "toolcall1",
        "type": "function",
      },
    ],
  },
  {
    "content": "7",
    "name": "CalculatorTool",
    "role": "tool",
    "status": "success",
    "toolCallId": "toolcall1",
  },
  {
    "content": "The CalculatorTool tool said: 7",
    "role": "assistant",
  },
]
`;
